TypeError: dist must be a Distribution instance

### Description

when i run tests with `nox -s test-3.9 -- -n auto tests/functional/test_bad_url.py  -v`

it is occurred an error, and can not passed. a lot of errors occurred when running the full number of test cases. i seek the answer on the internet, and some man say the reason is that setuptools is too new.

`my setuptools is : 59.4.0`


### Expected behavior

it should be running test `PASSED`

### pip version

21.3.1

### Python version

3.9.9

### OS

x86_64 linux

### How to Reproduce

nox -s test-3.9 -- -n auto tests/functional/test_bad_url.py  -v

### Output

```shell
nox > Running session test-3.9
nox > Re-using existing virtual environment at .nox/test-3-9.
nox > Re-using existing common-wheels at tests/data/common_wheels.
nox > python setup.py sdist --formats=zip --dist-dir /root/pip/.nox/test-3-9/sdist
nox > python tools/tox_pip.py install /root/pip/.nox/test-3-9/sdist/pip-21.3.1.zip
nox > python tools/tox_pip.py install -r tests/requirements.txt
nox > pytest -n auto tests/functional/test_bad_url.py -v
=============================================== test session starts ================================================
platform linux -- Python 3.9.9, pytest-7.1.2, pluggy-1.0.0 -- /root/pip/.nox/test-3-9/bin/python
cachedir: .pytest_cache
rootdir: /root/pip, configfile: setup.cfg
plugins: forked-1.4.0, rerunfailures-10.2, xdist-2.5.0, cov-3.0.0
[gw0] linux Python 3.9.9 cwd: /root/pip
[gw1] linux Python 3.9.9 cwd: /root/pip
[gw2] linux Python 3.9.9 cwd: /root/pip
[gw3] linux Python 3.9.9 cwd: /root/pip
[gw0] Python 3.9.9 (main, May 16 2022, 00:00:00)  -- [GCC 10.3.1]
[gw1] Python 3.9.9 (main, May 16 2022, 00:00:00)  -- [GCC 10.3.1]
[gw2] Python 3.9.9 (main, May 16 2022, 00:00:00)  -- [GCC 10.3.1]
[gw3] Python 3.9.9 (main, May 16 2022, 00:00:00)  -- [GCC 10.3.1]
gw0 [1] / gw1 [1] / gw2 [1] / gw3 [1]
scheduling tests via LoadScheduling

tests/functional/test_bad_url.py::test_filenotfound_error_message 
[gw0] [100%] ERROR tests/functional/test_bad_url.py::test_filenotfound_error_message 

====================================================== ERRORS ======================================================
________________________________ ERROR at setup of test_filenotfound_error_message _________________________________
[gw0] linux -- Python 3.9.9 /root/pip/.nox/test-3-9/bin/python

tmpdir_factory = TempdirFactory(_tmppath_factory=TempPathFactory(_given_basetemp=PosixPath('/tmp/pytest-of-root/pytest-43/popen-gw0'), ...luggy._tracing.TagTracerSub object at 0x7f64540d04f0>, _basetemp=PosixPath('/tmp/pytest-of-root/pytest-43/popen-gw0')))
common_wheels = Path('/root/pip/tests/data/common_wheels')

    @pytest.fixture(scope="session")
    def setuptools_install(
        tmpdir_factory: pytest.TempdirFactory, common_wheels: Path
    ) -> Path:
>       return _common_wheel_editable_install(tmpdir_factory, common_wheels, "setuptools")

tests/conftest.py:335: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/conftest.py:325: in _common_wheel_editable_install
    Wheel(wheel_candidates[0]).install_as_egg(install_dir)
.nox/test-3-9/lib/python3.9/site-packages/setuptools/wheel.py:96: in install_as_egg
    self._install_as_egg(destination_eggdir, zf)
.nox/test-3-9/lib/python3.9/site-packages/setuptools/wheel.py:104: in _install_as_egg
    self._convert_metadata(zf, destination_eggdir, dist_info, egg_info)
.nox/test-3-9/lib/python3.9/site-packages/setuptools/wheel.py:164: in _convert_metadata
    setup_dist.get_command_obj('egg_info'),
.nox/test-3-9/lib/python3.9/site-packages/setuptools/_distutils/dist.py:860: in get_command_obj
    cmd_obj = self.command_obj[command] = klass(self)
.nox/test-3-9/lib/python3.9/site-packages/setuptools/__init__.py:106: in __init__
    super().__init__(dist)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <setuptools.command.egg_info.egg_info object at 0x7f644fba3ee0>
dist = <setuptools.dist.Distribution object at 0x7f644fb859d0>

    def __init__(self, dist):
        """Create and initialize a new Command object.  Most importantly,
        invokes the 'initialize_options()' method, which is the real
        initializer and depends on the actual command being
        instantiated.
        """
        # late import because of mutual dependence between these classes
        from distutils.dist import Distribution
    
        if not isinstance(dist, Distribution):
>           raise TypeError("dist must be a Distribution instance")
E           TypeError: dist must be a Distribution instance

.nox/test-3-9/lib/python3.9/site-packages/setuptools/_distutils/cmd.py:57: TypeError
================================================= warnings summary =================================================
.nox/test-3-9/lib/python3.9/site-packages/_distutils_hack/__init__.py:33
.nox/test-3-9/lib/python3.9/site-packages/_distutils_hack/__init__.py:33
.nox/test-3-9/lib/python3.9/site-packages/_distutils_hack/__init__.py:33
.nox/test-3-9/lib/python3.9/site-packages/_distutils_hack/__init__.py:33
.nox/test-3-9/lib/python3.9/site-packages/_distutils_hack/__init__.py:33
  /root/pip/.nox/test-3-9/lib/python3.9/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
    warnings.warn("Setuptools is replacing distutils.")

.nox/test-3-9/lib/python3.9/site-packages/pip/_vendor/packaging/version.py:111: 540 warnings
  /root/pip/.nox/test-3-9/lib/python3.9/site-packages/pip/_vendor/packaging/version.py:111: DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================= short test summary info ==============================================
ERROR tests/functional/test_bad_url.py::test_filenotfound_error_message - TypeError: dist must be a Distribution ...
========================================== 545 warnings, 1 error in 2.97s ==========================================
nox > Command pytest -n auto tests/functional/test_bad_url.py -v failed with exit code 1
nox > Session test-3.9 failed.
```


### Code of Conduct

- [X] I agree to follow the [PSF Code of Conduct](https://www.python.org/psf/conduct/).

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -29,7 +29,9 @@ from _pytest.config import Config
 # Parser will be available from the public API in pytest >= 7.0.0:
 # https://github.com/pytest-dev/pytest/commit/538b5c24999e9ebb4fab43faabc8bcc28737bcdf
 from _pytest.config.argparsing import Parser
-from setuptools.wheel import Wheel
+from installer import install
+from installer.destinations import SchemeDictionaryDestination
+from installer.sources import WheelFile
 
 from pip._internal.cli.main import main as pip_entry_point
 from pip._internal.locations import _USE_SYSCONFIG
@@ -364,10 +366,29 @@ def _common_wheel_editable_install(
     wheel_candidates = list(common_wheels.glob(f"{package}-*.whl"))
     assert len(wheel_candidates) == 1, wheel_candidates
     install_dir = tmpdir_factory.mktemp(package) / "install"
-    Wheel(wheel_candidates[0]).install_as_egg(install_dir)
-    (install_dir / "EGG-INFO").rename(install_dir / f"{package}.egg-info")
-    assert compileall.compile_dir(str(install_dir), quiet=1)
-    return install_dir
+    lib_install_dir = install_dir / "lib"
+    bin_install_dir = install_dir / "bin"
+    with WheelFile.open(wheel_candidates[0]) as source:
+        install(
+            source,
+            SchemeDictionaryDestination(
+                {
+                    "purelib": os.fspath(lib_install_dir),
+                    "platlib": os.fspath(lib_install_dir),
+                    "scripts": os.fspath(bin_install_dir),
+                },
+                interpreter=sys.executable,
+                script_kind="posix",
+            ),
+            additional_metadata={},
+        )
+    # The scripts are not necessary for our use cases, and they would be installed with
+    # the wrong interpreter, so remove them.
+    # TODO consider a refactoring by adding a install_from_wheel(path) method
+    # to the virtualenv fixture.
+    if bin_install_dir.exists():
+        shutil.rmtree(bin_install_dir)
+    return lib_install_dir
 
 
 @pytest.fixture(scope="session")
@@ -389,13 +410,12 @@ def coverage_install(
     return _common_wheel_editable_install(tmpdir_factory, common_wheels, "coverage")
 
 
-def install_egg_link(
-    venv: VirtualEnvironment, project_name: str, egg_info_dir: Path
+def install_pth_link(
+    venv: VirtualEnvironment, project_name: str, lib_dir: Path
 ) -> None:
-    with open(venv.site / "easy-install.pth", "a") as fp:
-        fp.write(str(egg_info_dir.resolve()) + "\n")
-    with open(venv.site / (project_name + ".egg-link"), "w") as fp:
-        fp.write(str(egg_info_dir) + "\n.")
+    venv.site.joinpath(f"_pip_testsuite_{project_name}.pth").write_text(
+        str(lib_dir.resolve()), encoding="utf-8"
+    )
 
 
 @pytest.fixture(scope="session")
@@ -418,7 +438,7 @@ def virtualenv_template(
     venv = VirtualEnvironment(tmpdir.joinpath("venv_orig"), venv_type=venv_type)
 
     # Install setuptools and pip.
-    install_egg_link(venv, "setuptools", setuptools_install)
+    install_pth_link(venv, "setuptools", setuptools_install)
     pip_editable = tmpdir_factory.mktemp("pip") / "pip"
     shutil.copytree(pip_src, pip_editable, symlinks=True)
     # noxfile.py is Python 3 only
@@ -433,7 +453,7 @@ def virtualenv_template(
 
     # Install coverage and pth file for executing it in any spawned processes
     # in this virtual environment.
-    install_egg_link(venv, "coverage", coverage_install)
+    install_pth_link(venv, "coverage", coverage_install)
     # zz prefix ensures the file is after easy-install.pth.
     with open(venv.site / "zz-coverage-helper.pth", "a") as f:
         f.write("import coverage; coverage.process_startup()")
@@ -481,7 +501,7 @@ def virtualenv(
 
 @pytest.fixture
 def with_wheel(virtualenv: VirtualEnvironment, wheel_install: Path) -> None:
-    install_egg_link(virtualenv, "wheel", wheel_install)
+    install_pth_link(virtualenv, "wheel", wheel_install)
 
 
 class ScriptFactory(Protocol):

--- a/tests/functional/test_list.py
+++ b/tests/functional/test_list.py
@@ -109,7 +109,7 @@ def test_local_columns_flag(simple_script: PipTestEnvironment) -> None:
     assert "Package" in result.stdout
     assert "Version" in result.stdout
     assert "simple (1.0)" not in result.stdout
-    assert "simple     1.0" in result.stdout, str(result)
+    assert "simple  1.0" in result.stdout, str(result)
 
 
 def test_multiple_exclude_and_normalization(

--- a/tests/requirements.txt
+++ b/tests/requirements.txt
@@ -1,5 +1,6 @@
 cryptography
 freezegun
+installer
 pytest
 pytest-cov
 pytest-rerunfailures

